075-auto-rebalancing-cluster-scaling.md

```diff
@@ -0,0 +1,368 @@
+     * the `autoRebalancing()` method detects that a scaling up was requested because of `KafkaCluster.addedNodes()` returning a non empty list with the brokers' IDs to be added to the cluster. Because the `Kafka.status.cruiseControl.autoRebalance.state` doesn't exist, it's a new operation so the state is updated as `ScalingUp` and the `add-brokers` with the brokers' IDs. Of course, it doesn't trigger any rebalance because the scaling up is going on asynchronously.
+2. From the **2nd** to the **(N-1)th** reconciliation (scaling up is going on, checks when it's done):
+     * the `podSet()` does nothing because no new brokers were added this time.
+     * the `autoRebalancing()` evaluates the `Kafka.status.cruiseControl.autoRebalance.state: ScalingUp`, so a scaling up is running, and it has to check if it's finished or not. This can be achieved looking at the involved `StrimziPodSet`(s) and comparing `readyPods` and `pods` fields in the corresponding status. If they differ, the scaling up is still running. It should not take long time but it could go across multiple reconciliation, so this check repeats.
+3. On the **Nth** reconciliation (scaling up is done, rebalancing can start):
+     * the `podSet()` does nothing because no new brokers are added this time.
+     * the `autoRebalancing()` evaluates the `Kafka.status.cruiseControl.autoRebalance.state: ScalingUp`, so a scaling up is running, but finally looking at the involved `StrimziPodSet`(s), it gets `readyPods` equals to `pods` so the scaling up is done. A new `KafkaRebalance` custom resource is created and the `Kafka.status.cruiseControl.autoRebalance.state` is updated with the corresponding state (should be `ProposalPending`). The `KafkaRebalanceAssemblyOperator` takes care of handling the rebalancing.
+4. From the **(N+1)th** to the **(M-1)th** reconciliation (rebalancing is going on, checks when it's done):
+     * the `podSet()` does nothing because no new brokers are added this time. 
+     * the `autoRebalancing()` has to check the status of the rebalancing by looking at the corresponding `KafkaRebalance` custom resource status which will move through `ProposalPending`, `ProposalReady` to `Rebalancing` (we are assuming auto-approval). The rebalancing status is reflected into the `Kafka.status.cruiseControl.autoRebalance.state` field.
+5. On the **Mth** reconciliation (rebalancing id done):
+     * the `podSet()` does nothing because no new brokers are added this time.
+     * the `autoRebalancing()` detects that the `KafkaRebalance` custom resource status reports the `Ready` state (so the rebalancing is completed).
+
+During the above flow, errors can happen after the scaling up:
+
+* There is an error when getting the proposal.
+* After auto-approving the proposal, there is an error when running the rebalance.
```

Comment:
Actually, the current behaviour we have in the `KafkaRebalanceAssemblyOperator` is just returning a ... succeeded Future when a `KafkaRebalance` resource is deleted. It means, that any task started on CC will just continue without any feedback to the user. A proposal will be still calculated or a rebalancing will just continue.
From an auto-rebalancing perspective it means:

* after the scaling up, the autorebalancing will just proceed but no feedback for the user :-(
* before scaling down, the autorebancing will just proceed with no feedback for the user but when it ends the scale down should anyway happen because the scale-down check will detect no partitions are hosted anymore on removed brokers

In general, what we have today is not that great but I am also not sure if the correct behaviour should be stopping the current task, even because on `KafkaRebalance` deletion we don't have the user task ID anymore to ask for stopping, unless we start to track it in memory somehow. But this issue needs a different discussion and maybe a totally new proposal, unrelated to auto-rebalancing.

Reply:
> In general, what we have today is not that great but I am also not sure if the correct behaviour should be stopping the current task, even because on KafkaRebalance deletion we don't have the user task ID anymore to ask for stopping, unless we start to track it in memory somehow. But this issue needs a different discussion and maybe a totally new proposal, unrelated to auto-rebalancing.

I think stopping it would make some sense. But you cannot do it reliably without having a finalizer. So it would be a lot of work and problems with a little benefit I guess?

That said, if someone / something deletes the KafkaRebalance resource ... I do not think the KafkaReconciler can just rely on it being somewhere in progress, or? It would need to handle it by recreating it? Or it would need to raise this to the user and give the user some kind of remediation path?

Reply:
Recreating a correct `KafkaRebalance` (manually or automatically) is not that simple as well as having a straightforward remediation path. The `KafkaRebalance` status holds the state of the task (proposal pending, ready, rebalancing, ...) + the user task ID. When you delete the resource, you lose all this information.
The current `KafkaRebalanceAssemblyOperator` doesn't have any interaction with CC API unless there is a `KafkaRebalance` resource to reconcile. Rebuilding a correct `KafkaRebalace` would mean checking the CC API and reconstructing the status of a possible (missing) `KafkaRebalance` resources from there. Having a finalizer blocking the deletion would help I guess. The finalizer would be deleted when we are sure the `KafkaRebalance` reached a stable state (which could be proposal ready or ready, so when rebalancing is done).
I will open a different GitHub issue for this problem where we can discuss, because unrelated to the autorebalancing anyway.

Reply:
Yeah, I guess having a finalizer set on the KafkaRebalances we create directly would make sense and should help to protect things.

Reply:
You can simply create a duplicated KR which will essentially result in a noop if the previously created KR was completed successfully. This is not a big deal for CC.

Reply:
But you don't know nothing about in which state was the previous KR so you don't know if it was completed successfully or it's still running. When you delete a custom resource, you get `null` and not the deleted KR. So you can just create a new KR which would start a new task for a proposal and this will go into the CC tasks queue. In the end, you are not "tracking" the previous operation (related to the deleted KR), you are tracking a completely new operation.
Also, the automatic creation is something ok-ish for auto-rebalancing where the operator is already creating something for you, but not for the manual rebalancing. The user creates the KR, the operator should not re-create something created and deleted (by mistake) by the user.
I do still think that finalizer is the way to go avoiding the deletion until it's safe, unless forced (so the user knows what they are doing).

Reply:
> But you don't know nothing about in which state was the previous KR so you don't know if it was completed successfully or it's still running.

Exactly, so you create a twin KR that would be completed almost immediately if the previous one was completed successfully, or will be executed if not. This would be only in case auto-rebalance is on. Anyway, I think both approaches are valid, so it's your choice which strategy you prefer.

Reply:
> Yeah, I guess having a finalizer set on the KafkaRebalances we create directly would make sense and should help to protect things.

Just to be clear -> my idea was that when you create the internal rebalance from KafkaRecocniler you would add a finalizer to it in the KafkaReconciler. Not that the KafkaRebalanceAssemblyOperator would add finalizers to the rebalance resources created by the users.

Reply:
I don't see it as feasible because it's the `KafkaRebalanceAssemblyOperator` to know when it's the time to add/remove the finalizer. See the proposal for that https://github.com/strimzi/proposals/pull/123
I think the approach taken by the TO is the good one here. The finalizer is added by the TO to the `KafkaTopic`.

Reply:
Sorry, but I do not think the proposal does not help you here with anything. our concern here should be someone deleting the proposal the KafkaRecocniler created by mistake (e.g. not knowing what it is). You prevent that by setting the finalizer at creation by KafkaRecocniler and remove it in KafkaRecocniler when deleting it. Anything else will not help you because users will still come and delete it while getting a proposal.

Reply:
> For this specific issue, this usage of the finalizer could make sense. Maybe I misunderstood. But then the issue I opened (which was not related to it) didn't need a proposal as you suggested but just being closed, or?

That is why I tried (unsuccessfully) to stop you at the issue. A finalizer set and managed by the KafkaRecocniler for the auto-rebalancing should be mentioned in this proposal, but surely doesn't need a separate one.